patoolib/__init__.py

```diff
@@ -558,7 +620,25 @@ def get_archive_cmdlist_func (program, command, format):
         raise util.PatoolError(msg)
     # get archive handler function (eg. patoolib.programs.star.extract_tar)
     try:
-        return getattr(module, '%s_%s' % (command, format))
+        archive_cmdlist_func = getattr(module, '%s_%s' % (command, format))
+        def check_for_password_before_cmdlist_func_call(*args, **kwargs):
+            """ If password is None, or not set, run command as usual.
+            If password is set, but can't be accepted raise appropriate
+            message.
+            """
+            util.log_info("... cmdlist_func = %s %s" % (archive_cmdlist_func, ''))
+            util.log_info("... kwargs=%s args=%s" % (kwargs, args))
+            if 'password' in kwargs and kwargs['password'] is None:
+                kwargs.pop('password')
+            if 'password' not in kwargs:
+                return archive_cmdlist_func(*args, **kwargs)
+            else:
+                try:
+                    return archive_cmdlist_func(*args, **kwargs)
+                except TypeError as msg:
+                    if "unexpected keyword argument 'password'" in str(msg):
```

Comment:
this looks suboptimal (sure it would never be `"password"` instead of `'password'` and message itself does not change?) -- if we are to rely on keyword arg to be known, may be some inspection to be done instead of catching/analyzing TypeError?  and don't we already have that information in NoPasswordSupportArchivePrograms?

Reply:
Yes, I thought about it, and not sure about better solution. I tried to realize it in pythonic way: try something, than catch error. NoPasswordSupportArchivePrograms was made as double of ArchivePrograms, to get subset of it, which support password in archives. This way you can just add new archive programs to AchivePrograms, set password argument and it will work.

I can make, instead, PasswordSupportArchivePrograms, which will realize exact that subset. Minus here is that anyone, who would add some program with password support, will need to add it to both dicts: ArchivePrograms and PasswordSupportArchivePrograms. Will it be better than current realization?

Reply:
what about `'password' in inspect.signature(archive_cmdlist_func).parameters` ?

Reply:
Yes, I think it will be much better. I will change it now.

Reply:
Changed